File: game/main_folder/game.py
import graphics
import other_calculations


class Board:
    def __init__(self, players, board, limit, game_graphics):
        self.players = players
        self.board = board
        self.limit = limit
        self.turn = 0
        self.game_graphics = game_graphics

    # check whose turn it is
    def get_turn(self):
        if self.turn == self.limit:
            self.turn = 0
        player = self.players[self.turn]
        return player

    def move(self, player, movement):
        # piece movement
        print(movement)
        if movement["type"] == "piece move":
            if movement["movement"] == "forward":
                player.pawn.position += player.direction*34
            elif movement["movement"] == "backward":
                player.pawn.position += -1*player.direction*34
            elif movement["movement"] == "right":
                player.pawn.position += player.direction*2
            elif movement["movement"] == "left":
                player.pawn.position -= player.direction*2
            elif movement["movement"] == "jump forward":
                player.pawn.position += player.direction*68
            elif movement["movement"] == "jump backward":
                player.pawn.position -= player.direction*68
            elif movement["movement"] == "jump right":
                player.pawn.position += player.direction*4
            elif movement["movement"] == "jump left":
                player.pawn.position -= player.direction*4
            elif movement["movement"] == "right top diagonal":
                player.pawn.position += player.direction*36
            elif movement["movement"] == "left top diagonal":
                player.pawn.position += player.direction*32
            elif movement["movement"] == "left down diagonal":
                player.pawn.position -= player.direction*36
            elif movement["movement"] == "right down diagonal":
                player.pawn.position -= player.direction*32
            if player.direction == 1 and player.pawn.position > 17*16:
                self.winner(player, movement)
                return True
            elif player.direction == -1 and player.pawn.position < 18:
                self.winner(player, movement)
                return True

        # fence placement
        if movement["type"] == "fence placement":
            for pos in movement["movement"]:
                self.board[pos-1] = 1
            other_player = None
            if player.board.turn == 1:
                other_player = player.board.players[0]
            else:
                other_player = player.board.players[1]
            if not other_calculations.is_path(player.pawn.position, self.board, [], player.direction) or not other_calculations.is_path(other_player.pawn.position, self.board, [], other_player.direction):
                for pos in movement["movement"]:
                    self.board[pos - 1] = 0
                player.send({"command": "invalid reply", "problem": "invalid movement"})
                return False
        # moving to the next person
        self.turn += 1
        turn = self.get_turn()
        print(player.pawn.position)
        self.print_board()
        self.send_all({"command": "moved", "player_code": player.client.code, "movement": movement})
        self.send_all({"command": "turn to move", "turn": turn.client.code})

    def add_player(self, player):
        player_direction = 0

        # setting up the player's pawn position
        if len(self.players) == 0:
            player.pawn.position = 9
            player_direction = 1
        if len(self.players) == 1:
            player.pawn.position = 281
            player_direction = -1

        # setting up player
        self.players.append(player)
        player.setup(self.board, player_direction)
        # print("how many times")

        # starting the game
        if len(self.players) >= self.limit:
            self.start()

    def start(self):
        self.game_graphics.storage["status"] = "playing"
        nicknames = [player.nickname for player in self.players]
        self.send_all({"command": "begin game", "players nickname": nicknames})
        turn = self.get_turn()
        self.send_all({"command": "turn to move", "turn": turn.client.code})

    def send_all(self, msg):
        for player in self.players:
            player.send(msg)

    def winner(self, winner, movement):
        self.send_all({"command": "moved", "player_code": winner.client.code, "movement": movement})
        self.game_graphics.storage["status"] = "over"
        self.send_all({"command": "winner", "winner": winner.client.code})
        graphics.game_graphics_list.pop(self.game_graphics)

    def print_board(self):
        counter = 0
        for unit in self.board:
            counter += 1
            print(unit, end=' ')
            if counter == 17:
                print()
                counter = 0


class Game:
    def __init__(self, board):
        self.board = board


class Pawn:
    def __init__(self, player, position):
        self.player = player
        self.position = position


class Player:
    def __init__(self, pawn, board, client, nickname):
        self.pawn = pawn
        self.board = board
        self.client = client
        self.nickname = nickname
        self.direction = None

    def send(self, msg):
        self.client.send(msg)

    def setup(self, board, direction):
        self.direction = direction
        msg = {"command": "setup", "board": board, "nickname": self.nickname, "direction": self.direction, "client code": self.client.code}
        self.send(msg)


def new_game(game_graphics):
    board_list = [0 for _ in range((9*2-1)**2)]
    board = Board([], board_list, 2, game_graphics)
    game_graphics.storage["board"] = board
    game_graphics.storage["players"] = []
    game_graphics.storage["status"] = "waiting for players"


def new_player(game_graphics, client, nickname):
    pawn = Pawn(None, 0)
    player = Player(pawn, game_graphics.storage["board"], client, nickname)
    pawn.player = player
    game_graphics.storage["board"].add_player(player)
    game_graphics.storage["players"].append(player)
    return player

File: game/networking/static/game_tools/game.js
var socket = io();

window.tile_height = 50;
window.screen_width = 9*tile_height+8*tile_height*0.2;
window.screen_height = 9*tile_height+8*tile_height*0.2;

const canvas = document.querySelector('canvas');
const ctx = canvas.getContext('2d');
canvas.width = window.screen_width;
canvas.height = window.screen_height;


window.client_code = 0;
window.this_player = new Player("", 0, 1, null);
window.this_opponent = new Player("", 0, -1, null);
window.fences = [];

window.fence_mode = false;
window.horizontal = true;
window.turn = false;

window.winner = null;
window.game_status = "starting ig"


// clearing the screen
function clear(){
    // clear screen 
    ctx.clearRect(0, 0, canvas.width, canvas.height);
}


// NETWORKING

// sending messages
function send(entery){
    socket.send(JSON.stringify(entery));
}

// on server commands
socket.on('message', function(msg) {
    var msg = JSON.parse(msg);
    console.log(msg);
    if (msg["command"] == "setup"){
        // storing the information from server's message
        window.client_code = msg["client code"];
        window.this_player.direction = msg["direction"];
        if (window.this_player.direction == 1){
            // setting up this player's position
            window.this_player.position = 9;

            // setting up the opponent's position and direction
            window.this_opponent.position = 281;
            window.this_opponent.direction = -1;
        }
        else if (window.this_player.direction == -1){
            // setting up this player's position 
            window.this_player.position = 281;

            // setting up the opponent's position and direction
            window.this_opponent.position = 9;
            window.this_opponent.direction = 1;
        }

        window.board = new Board(window.this_player, window.this_opponent, window.fences, [], ctx);
        window.this_player.board = window.board;
        window.this_opponent.board = window.board;
        window.board.board = msg["board"];
        window.board.create_board_shapes();
        
        window.board.draw_board();

        document.getElementById("game_status").innerHTML = "STATUS: Waiting for the opponent impatiently :\\";

        // sending the needed information to the server
        send({"command":"setup reply", "nickname": nickname});
    }
    else if (msg["command"] == "begin game"){
        // setting up the opponent's nickname
        if (msg["players nickname"][0] == window.this_player.nickname){
            window.this_opponent.nickname = msg["players nickname"][1];
        }
        else {
            window.this_opponent.nickname = msg["players nickname"][0];
        }
        document.getElementById("game_status").innerHTML = "STATUS: Playing";
    }
    else if (msg["command"] == "turn to move"){
        // setting the turn
        if (msg["turn"] == window.client_code){
            window.turn = true;
            document.getElementById("turn").innerHTML = "TURN: You";
        }
        else{
            window.turn = false;
            document.getElementById("turn").innerHTML = "TURN: Opponent";
        }
    }
    else if (msg["command"] == "moved"){
        // moving this player
        if (msg["player_code"] == window.client_code){
            window.this_player.move(msg["movement"]["type"], msg["movement"]["movement"]);
            if (msg["movement"]["type"] == "fence placement"){
                document.getElementById("fences_left").innerHTML = "FENCES LEFT: "+window.this_player.fences;
            }
        }
        // moving the opponent
        else{
            window.this_opponent.move(msg["movement"]["type"], msg["movement"]["movement"]);
        }
        window.board.draw_board();
    }
    else if (msg["command"] == "invalid reply"){
        window.turn = true;
        console.log(msg["problem"]);
        if (msg["problem"] == "invalid movement"){
            window.turn = true;
        }
    }
    else if (msg["command"] == "winner"){
        window.turn = false;
        if (msg["winner"] == window.client_code){
            window.winner = window.this_player;
            document.getElementById("game_status").innerHTML = "STATUS: You won :)";
            window.game_status = "over";
        }
        else {
            window.winner = window.this_opponent;
            document.getElementById("game_status").innerHTML = "STATUS: You lost :(";
            window.game_status = "over";
        }
    }

});


// Event listeners
canvas.addEventListener("mousemove", function(e) {
    var cRect = canvas.getBoundingClientRect();        // Gets CSS pos, and width/height
    var x = Math.round(e.clientX - cRect.left);  // Subtract the 'left' of the canvas
    var y = Math.round(e.clientY - cRect.top);   // from the X/Y positions to make
    if (window.turn){
        var x_position = Math.ceil(x/(window.tile_height+window.tile_height*0.2));
        var y_position = Math.ceil(y/(window.tile_height+window.tile_height*0.2));
        var position = x_position+x_position-1+17*(y_position+y_position-2);
        if (window.fence_mode){
            var poses = [];
            if (position%17 == 0 && position/17 > 16){
                if (window.horizontal){
                    poses.push(position-19);                
                    poses.push(position-18);
                    poses.push(position-17);
                }
                else{
                    poses.push(position-1);
                    poses.push(position-18);
                    poses.push(position-35);
                }
            }
            else if (position%17 == 0){  
                if (window.horizontal){
                    poses.push(position+15);                
                    poses.push(position+16);
                    poses.push(position+17);
                }
                else{
                    poses.push(position-1);
                    poses.push(position+16);
                    poses.push(position+33);
                }
                
            }
            else if (position/17 > 16){
                if (window.horizontal){
                    poses.push(position-17);
                    poses.push(position-16);
                    poses.push(position-15);
                }
                else{
                    poses.push(position-33);
                    poses.push(position-16);
                    poses.push(position+1);
                }
            }
            else{
                if (window.horizontal){
                    poses.push(position+17);
                    poses.push(position+18);
                    poses.push(position+19);
                }
                else{
                    poses.push(position+1);
                    poses.push(position+18);
                    poses.push(position+35);
                }
            }
            if (window.board.option_fence_pos != poses){
                window.board.draw_option_fence = true;
                var color = [255, 0, 0];
                if (window.this_player.check_movement("fence placement", poses)){
                    color = [0, 255, 0];
                }
                window.board.option_fence_pos = poses;
                window.board.option_fence_shape.color = rgbToColor(color);
                window.board.draw_board();
            }
        }
        else if (window.board.option_player_pos != position && !window.fence_mode){
            window.board.draw_option_player = true;
            window.board.option_player_pos = position;
            var movement = "";
            var color = [0, 255, 0];
            if (position == window.this_player.position + window.this_player.direction*34){
                movement = "forward";
            }
            else if (position == window.this_player.position - window.this_player.direction*34){
                movement = "backward";
            }
            else if (position == window.this_player.position + window.this_player.direction*2){
                movement = "right";
            }
            else if (position == window.this_player.position - window.this_player.direction*2){
                movement = "left";
            }
            else if (position == window.this_player.position + window.this_player.direction*68){
                movement = "jump forward";
            }
            else if (position == window.this_player.position - window.this_player.direction*68){
                movement = "jump backward";
            }
            else if (position == window.this_player.position + window.this_player.direction*4){
                movement = "jump right";
            }
            else if (position == window.this_player.position - window.this_player.direction*4){
                movement = "jump left";
            }
            else if (position == window.this_player.position + window.this_player.direction*36){
                movement = "right top diagonal";
            }
            else if (position == window.this_player.position + window.this_player.direction*32){
                movement = "left top diagonal";
            }
            else if (position == window.this_player.position - window.this_player.direction*36){
                movement = "left down diagonal";
            }
            else if (position == window.this_player.position - window.this_player.direction*32){
                movement = "right down diagonal";
            }
            else {
                color = [255, 0, 0];
            }
            // console.log(window.this_player.position + window.this_player.direction*35);
            // console.log(position);
            // console.log(movement);
            if (!window.this_player.check_movement("piece move", movement)){
                color = [255, 0, 0];
            }
            window.board.option_player_shape.color = rgbToColor(color);
            window.board.draw_board();
        }
        // console.log(position);
    }
});

canvas.addEventListener("mouseup", function(e){
    var cRect = canvas.getBoundingClientRect();        
    var x = Math.round(e.clientX - cRect.left);  
    var y = Math.round(e.clientY - cRect.top);   
    if (window.turn){
        var x_position = Math.ceil(x/(window.tile_height+window.tile_height*0.2));
        var y_position = Math.ceil(y/(window.tile_height+window.tile_height*0.2));
        var position = x_position+x_position-1+17*(y_position+y_position-2);
        if (window.fence_mode){
            var poses = [];
            if (position%17 == 0 && position/17 > 16){
                if (window.horizontal){
                    poses.push(position-19);                
                    poses.push(position-18);
                    poses.push(position-17);
                }
                else{
                    poses.push(position-1);
                    poses.push(position-18);
                    poses.push(position-35);
                }
            }
            else if (position%17 == 0){  
                if (window.horizontal){
                    poses.push(position+15);                
                    poses.push(position+16);
                    poses.push(position+17);
                }
                else{
                    poses.push(position-1);
                    poses.push(position+16);
                    poses.push(position+33);
                }
                
            }
            else if (position/17 > 16){
                if (window.horizontal){
                    poses.push(position-17);
                    poses.push(position-16);
                    poses.push(position-15);
                }
                else{
                    poses.push(position-33);
                    poses.push(position-16);
                    poses.push(position+1);
                }
            }
            else{
                if (window.horizontal){
                    poses.push(position+17);
                    poses.push(position+18);
                    poses.push(position+19);
                }
                else{
                    poses.push(position+1);
                    poses.push(position+18);
                    poses.push(position+35);
                }
            }
            if (window.this_player.check_movement("fence placement", poses)){
                window.turn = false;
                send({"command": "moved", "movement":{"type": "fence placement", "movement": poses}});
                window.turn = false;
            }
            
        }
        else if (!window.fence_mode){
            var movement = "";
            if (position == window.this_player.position + window.this_player.direction*34){
                movement = "forward";
            }
            else if (position == window.this_player.position - window.this_player.direction*34){
                movement = "backward";
            }
            else if (position == window.this_player.position + window.this_player.direction*2){
                movement = "right";
            }
            else if (position == window.this_player.position - window.this_player.direction*2){
                movement = "left";
            }
            else if (position == window.this_player.position + window.this_player.direction*68){
                movement = "jump forward";
            }
            else if (position == window.this_player.position - window.this_player.direction*68){
                movement = "jump backward";
            }
            else if (position == window.this_player.position + window.this_player.direction*4){
                movement = "jump right";
            }
            else if (position == window.this_player.position - window.this_player.direction*4){
                movement = "jump left";
            }
            else if (position == window.this_player.position + window.this_player.direction*36){
                movement = "right top diagonal";
            }
            else if (position == window.this_player.position + window.this_player.direction*32){
                movement = "left top diagonal";
            }
            else if (position == window.this_player.position - window.this_player.direction*36){
                movement = "left down diagonal";
            }
            else if (position == window.this_player.position - window.this_player.direction*32){
                movement = "right down diagonal";
            }
            if (window.this_player.check_movement("piece move", movement)){
                window.turn = false;
                send({"command":"moved", "movement": {"type": "piece move", "movement": movement}});
            }
        }
        // console.log(position);
    }
})

document.addEventListener("keydown", function(e){
    // console.log(e.keyCode);
    if (e.keyCode == 70){
        window.fence_mode = !window.fence_mode;
        if (window.fence_mode){
            document.getElementById("fence_mode").innerHTML = "FENCE MODE: ON";
        }
        else{
            document.getElementById("fence_mode").innerHTML = "FENCE MODE: OFF";
        }
    }
    else if (e.keyCode == 72){
        window.horizontal = !horizontal;
    }
});


